### PR TITLE
[NPU] add npu support for conv3d and conv3d_grad

### PR DESCRIPTION
**Pull request by windstamp**
### PR types

New features
### PR changes

OPs
### Describe

add npu support for conv3d and conv3d_grad

### CODE REVIEW

File: paddle/fluid/operators/conv_op_npu.cc [hunk 1]
```diff
@@ -393,6 +393,236 @@ class NPUConvGradOpKernel : public framework::OpKernel<T> {
...
+                          "= [%s]",
+                          data_format));
+
+    PADDLE_ENFORCE_EQ(groups, 1, platform::errors::Unimplemented(
+                                     "the groups must be 1 in "
+                                     "the npu kernel of conv3d, but got groups "
+                                     "= [%d]",
+                                     groups));
+
+    const bool channel_last = data_format == "NDHWC";
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: 在413行 data_format 必须得是 "NCDHW"，这里就不需要增加对data_format == "NDHWC"的判断了？

File: paddle/fluid/operators/conv_op_npu.cc [hunk 2]
```diff
@@ -393,6 +393,236 @@ class NPUConvGradOpKernel : public framework::OpKernel<T> {
...
+
+    if (channel_last) {
+      input->set_layout(DataLayout::kNDHWC);
+      filter->set_layout(DataLayout::kNDHWC);
+      output->set_layout(DataLayout::kNDHWC);
+    } else {
+      input->set_layout(DataLayout::kNCDHW);
+      filter->set_layout(DataLayout::kNCDHW);
+      output->set_layout(DataLayout::kNCDHW);
+    }
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: 理论上这里的input和filter应该都是const Tensor，最好不要对他们的输入layout作改动，这里建议改成对input和filter的LAYOUT检查，如果和data_format值不一致就抛错

File: paddle/fluid/operators/conv_op_npu.cc [hunk 3]
```diff
@@ -393,6 +393,236 @@ class NPUConvGradOpKernel : public framework::OpKernel<T> {
...
+
+    // update padding and dilation
+    auto in_dims = input->dims();
+    auto filter_dims = filter->dims();
+    framework::DDim in_data_dims;
+    framework::DDim filter_data_dims;
+
+    if (channel_last) {
+      in_data_dims = framework::slice_ddim(in_dims, 1, in_dims.size() - 1);
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: 同上，这里 channel_last的情况应该不存在，因为PADDLE_ENFORCE_EQ(data_format, "NCDHW"？

File: paddle/fluid/operators/conv_op_npu.cc [hunk 4]
```diff
@@ -393,6 +393,236 @@ class NPUConvGradOpKernel : public framework::OpKernel<T> {
...
+
+    std::vector<int> strides_vec(5, 1);
+    std::vector<int> dilations_vec(5, 1);
+
+    Tensor input_tensor, output_tensor;
+
+    input_tensor.ShareDataWith(*input);
+    output_tensor.ShareDataWith(*output);
+
+    if (channel_last) {
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: 同上

File: paddle/fluid/operators/conv_op_npu.cc [hunk 5]
```diff
@@ -393,6 +393,236 @@ class NPUConvGradOpKernel : public framework::OpKernel<T> {
     }
   }
 };
+
+template <typename T>
+class NPUConv3dKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    Tensor* input = const_cast<Tensor*>(ctx.Input<Tensor>("Input"));
+    Tensor* filter = const_cast<Tensor*>(ctx.Input<Tensor>("Filter"));
```
Comment (qili93) [inline paddle/fluid/operators/conv_op_npu.cc]: Paddle代码中使用const_cast需要非常小心，这里理论上input/filter都不可以在Compute过程中被修改，建议改成对input和filter的LAYOUT检查，如果和data_format值不一致就抛错来提醒用户

File: paddle/fluid/operators/conv_op_npu.cc (hunk 1 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: Done, thanks.

开始想保留 data_format == "NDHWC" 的相关代码是，目前主要是由于 Ascend 底层抛错导致的，后续 Ascend 修复了可能直接把校验删除即可。不过先去除无效代码会更简洁，而且这些无效代码还导致了覆盖率 CI 失败。已删除这些无效代码。

File: paddle/fluid/operators/conv_op_npu.cc (hunk 2 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: 这个情况可能有点复杂，目前 Paddle 中有多处 DataLayout 定义（这里使用的定义（之前不包含 kNCDHW 等），以及 cuda 或 cudnn 相关的，rocm 或 miopen 相关的，后两者已经包含 kNCDHW）。这里的实现，从 Python 端设置的 "NCDHW" 等格式不会传入到 C++ 的 OP Kernel 中，如果不在这里主动设置一下，会导致 C++ 端等 data_format 并不正确，然后导致映射给 NPU 的数据格式不正确。

NCDHW 和 NDHWC 不是很清楚为什么之前没有在这里使用的 DataLayout 中定义。这里新增之后为了避免对其它模块的影响，没有做相关的处理，所以导致在 Python 层面的设置没有传递到 C++ 端。为了避免对其它模块的影响，我这里在 Kernel 中进行了处理。

但是这个方法可能也不是很合适，却掉 input 参数的 const 属性，确实也有可能存在风险。

File: paddle/fluid/operators/conv_op_npu.cc (hunk 3 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: Done, thanks.

File: paddle/fluid/operators/conv_op_npu.cc (hunk 4 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: Done, thanks.

File: paddle/fluid/operators/conv_op_npu.cc (hunk 5 above)
Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: 解释同上。

File: paddle/fluid/operators/conv_op_npu.cc (hunk 2 above)
Comment (qili93) [reply paddle/fluid/operators/conv_op_npu.cc]: 这里建议可以和OP负责人商量下，如果一定要修改也不可以修改只读的Input Tensor，否则对后续其他节点的OP输入会造成影响。举例，这里的同一个input tensor既是conv op的输入，也是模型网络中 batchnorm op的输入，假设conv在batchnorm之前运行然后在计算过程中修改了input tensor本身的layout信息，就会导致 batchnorm输入的Input tensor的layout信息也变掉了，就会导致计算方式不一致，原网络的精度就对不齐了。

如果一定要修改建议参考 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/batch_norm_op_npu.cc#L62 这里代码用ShareDataWith的方式，共享数据但是不是同一个tensor，这样能保留所有输入的const Tensor的layout等信息都不回在计算过程中被修改。

Comment (windstamp) [reply paddle/fluid/operators/conv_op_npu.cc]: 把 const_cast 去掉了。

目前根据上面的参考是这样实现的，
```
auto input_tensor =
        ctx.AllocateTmpTensor<T, NPUDeviceContext>(input->dims(), dev_ctx);
input_tensor.ShareDataWith(*input);
```

**Review by qili93 — APPROVED** [review]
LGTM
